src/main/java/com/mraof/minestuck/computer/editmode/ClientEditToolDrag.java

```diff
@@ -0,0 +1,447 @@
+import com.mraof.minestuck.block.machine.MachineBlock;
+import com.mraof.minestuck.network.EditmodeDragPacket;
+import com.mraof.minestuck.network.MSPacketHandler;
+import com.mraof.minestuck.player.ClientPlayerData;
+import com.mraof.minestuck.util.MSCapabilities;
+import com.mraof.minestuck.util.MSSoundEvents;
+import net.minecraft.client.Camera;
+import net.minecraft.client.KeyMapping;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.MultiBufferSource;
+import net.minecraft.client.renderer.RenderType;
+import net.minecraft.core.BlockPos;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.ClipContext;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.Material;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.BlockHitResult;
+import net.minecraft.world.phys.Vec3;
+import net.minecraft.world.phys.shapes.Shapes;
+import net.minecraft.world.phys.shapes.VoxelShape;
+import net.minecraftforge.api.distmarker.Dist;
+import net.minecraftforge.client.event.RenderLevelStageEvent;
+import net.minecraftforge.common.ForgeMod;
+import net.minecraftforge.event.TickEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.LogicalSide;
+import net.minecraftforge.fml.common.Mod;
+
+/** Class for handling the click-and-drag editmode tools (Revise and Recycle) on the client-side.
+ * (Based on code from the Minestuck Universe addon, with Cibernet's permission.)
+ * @see EditmodeDragPacket for the tool's server-sided block-placing code.
+ * @see ServerEditHandler for server-sided code that handles the sburb-cursor.
+ * @author Caldw3ll, Cibernet
+ */
+@Mod.EventBusSubscriber(modid = Minestuck.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE, value = Dist.CLIENT)
+public class ClientEditToolDrag
+{
+	
+	@SubscribeEvent
+	public static void onClientTick(TickEvent.ClientTickEvent event)
+	{
+		if(event.side == LogicalSide.CLIENT)
+		{
+			ClientEditToolDrag.doRecycleCode(event);
+			ClientEditToolDrag.doReviseCode(event);
+		}
+	}
+	
+	@SubscribeEvent
+	public static void renderWorld(RenderLevelStageEvent event)
+	{
+		ClientEditToolDrag.renderOutlines(event);
+	}
+	
+	/**
+	 * Resets the drag tool, and removes the server-cursor if the given edit tool is active.
+	 * @param targetTool The tool you want to check. Must be a drag tool (Revise or Recycle).
+	 * @param cap The current edit-tools capability.
+	 */
+	private static void cancelDrag(IEditTools.ToolMode targetTool, IEditTools cap)
+	{
+		if(!isValidDragTool(targetTool))
+			throw new IllegalArgumentException("targetTool in cancelDrag() must be a drag tool (Revise or Recycle)!");
+		
+		if(cap.getToolMode() == targetTool)
```

Comment:
Perhaps move this check to the if-statements outside the `cancelDrag()` functions? I think it should be seen as a condition to when `cancelDrag()` gets called (I don't think it should be a problem that `cap.resetDragTools();` no longer gets called when `cap.getToolMode() != targetTool` by moving this condition)